[KB] Customize Semgrep in pre-commit (#1889)
semgrep/semgrep-docs: docs/kb/integrations/customize-semgrep-precommit.md

---
description: Understand how to customize Semgrep's behavior when using it with pre-commit.
tags:
  - Extensions
---

# Customize Semgrep in pre-commit

Semgrep offers two standard pre-commit hooks [outlined here](/docs/extensions/overview/#pre-commit).

* The `semgrep` hook is optimized to run with a provided set of rules, either local rules or rules from the Semgrep Registry.

> * The `semgrep-ci` hook is optimized to run with the rules from your organization in the Semgrep AppSec Platform, without sending results to the platform, since results from pre-commit are temporary and not linked to the final code in the repository.

Author: Credit to pad on that I think!

> * The `semgrep-ci` hook is optimized to run with the rules from your organization in the Semgrep AppSec Platform, without sending results to the platform, since results from pre-commit are temporary and not linked to the final code in the repository.

Reviewer: wonder if it's worth calling out that this is near identical to what happens when semgrep runs in CI on a PR specifically?

Reviewer: Also, maybe we want to mention something about this requiring login?

Author: Plus on the login, minus on the "what happens when Semgrep runs on a PR" - it's actually significantly different from that since PR scans compare head and base, and pre-commit scans compare per-commit. We've had folks think that PR scans do per-commit analysis before and be confused that they don't, so I don't want to encourage the confusion.

You can also create your own Semgrep hooks for pre-commit if you have particular needs or preferences, such as if you want users to see the output of the Semgrep scan even if it passed, or if you want to limit the scan to a particular Semgrep product.

## Show the scan output

> By default, pre-commit does not show the scan output if the hook passes. If you're using the `semgrep-ci` hook, but want your developers to see the output even if non-blocking findings are found, you can set the `verbose` option to print the output. To limit output to findings only (no scan information or diagnostics), redirect `stderr` to `/dev/null`, or use `--quiet`.

Reviewer: what if I'm just using the semgrep hook, not semgrep-ci? Do I see the output then?

Author: This is just written as an example, but I'll make that clearer in the language.

> By default, pre-commit does not show the scan output if the hook passes. If you're using the `semgrep-ci` hook, but want your developers to see the output even if non-blocking findings are found, you can set the `verbose` option to print the output. To prevent the output from being too chatty, redirect `stderr` to `/dev/null` so that only the findings are printed, or use `--quiet`.

Reviewer 2: ```suggestion
By default, pre-commit does not show the scan output if the hook passes. If you're using the `semgrep-ci` hook, but want your developers to see the output even if non-blocking findings are found, you can set the `verbose` option to print the output. To prevent the output from being too chatty, redirect `stderr` to `/dev/null` so that only the findings are printed, or use `--quiet`.
```
Extremely minor nit, Instead of "too chatty" do we want to mention that verbose prints out errors too, then provide the redirection?
Something like "The `verbose` option logs error messages as well. You can reduce the logs by redirecting `stderr`..."

Author: That rephrase wouldn't be quite right here, since we print out all scan progress (diagnostic) information on stderr, not just errors. Findings are the only thing that goes to stdout. But I can choose a more neutral phrase for sure :)

Author: OK, updated :) Good call.

```yaml
repos:
- repo: https://github.com/semgrep/pre-commit
  rev: 'v1.101.0'
  hooks:
    - id: semgrep-verbose
      entry: semgrep
      args: ["ci", "--dry-run", "--baseline-commit", "HEAD" "2>/dev/null"]
      verbose: true
      pass_filenames: false
```

## Limit the scan to a particular product

Semgrep Secrets is an ideal product to run before commit, since it can help prevent secrets from ever making it into the Git history, even locally. To run only Secrets in pre-commit, add the product flag to the `args`:

```yaml
- repo: https://github.com/semgrep/pre-commit
  rev: 'v1.101.0'
  hooks:
    - id:  semgrep-secrets
      pass_filenames: false
      args: ["ci", "--dry-run", "--baseline-commit", "HEAD", "--secrets"]
```

## Customization tips

### Useful arguments

The provided hooks include useful arguments for Semgrep that you may want to include in your hooks as well.

> For example, for the `semgrep` hook that runs with a provided `--config`, the arguments also include:

Reviewer: we probably want --pro as a useful argument too

Author: This text is meant to cover "hey some useful arguments are used already, I'm going to explain them", not to introduce other arguments. But I can add an example or a note about `--pro` separately.

Reviewer: gotcha!

By default, pre-commit does not show the scan output if the hook passes. If you want your developers to see the output even if non-blocking findings are found, you can set the `verbose` option to print the output. To limit output to findings only (no scan information or diagnostics), redirect `stderr` to `/dev/null`, or use `--quiet`.

This example is based on the `semgrep-ci` hook, but a similar adjustment would also work with the `semgrep` hook.

```yaml
repos:
- repo: https://github.com/semgrep/pre-commit
  rev: 'v1.101.0'
  hooks:
    - id: semgrep-verbose
      entry: semgrep
      args: ["ci", "--dry-run", "--baseline-commit", "HEAD" "2>/dev/null"]
      verbose: true
      pass_filenames: false
```

## Limit the scan to a particular product

Semgrep Secrets is an ideal product to run before commit, since it can help prevent secrets from ever making it into the Git history, even locally. To run only Secrets in pre-commit, add the product flag to the `args`:

```yaml
- repo: https://github.com/semgrep/pre-commit
  rev: 'v1.101.0'
  hooks:
    - id:  semgrep-secrets
      pass_filenames: false
      args: ["ci", "--dry-run", "--baseline-commit", "HEAD", "--secrets"]
```

> ## Scan with Pro Engine

Reviewer 2: ```suggestion
## Scan with interfile analysis
```
From a marketing/product standpoint, we no longer use the term "Pro Engine" externally. (The only "Pro" product we have is "Pro rules".) We try to call it "interfile analysis" where possible.

Author: Unfortunately that wouldn't be right here, since diff-aware scans (scans with a baseline ref) don't run interfile analysis. Maybe just "Scan with Pro" then?

Reviewer 2: Hmm... When you say scan with Pro, this is the binary, and does NOT affect the Pro rules, right? Do we do interprocedural for diff scans? (my last knowledge was "this is in beta" but I could be wrong). We can say "cross-function analysis" if it's already in GA.

Author: yeah, diff scan is intrafile - the beta was for interfile, and we killed it :rip:

Author: I think this would apply to rules - let me check, though. The reason I think it will is that any token should unlock feature available to users of the platform, which includes pro rules.

Author: OK yeah - this also gave me a better idea of what to write here, based on what the CLI prints - if you're logged in, it says "Semgrep Code", and that might be a good choice for this purpose. Thoughts @s-santillan ?

Reviewer 2: Yes, we can use "Scan with Semgrep Code" or "Scan with Pro rules and cross-function analysis". If we go with "Semgrep Code" we can just mention what that entails (the Pro rules and cross-function analysis) in the `<p>`.

> The `semgrep-ci` hook requires the user to be logged in locally and runs with the engine configured in the organization, but the standard `semgrep` hook can also take advantage of local login to run with the Pro Engine.

Reviewer 2: ```suggestion
The `semgrep-ci` hook requires the user to be logged in locally and runs with the engine configured in the organization, but the standard `semgrep` hook can also take advantage of local login to run with interfile analysis.
```

Author: also wrong for the same reason, sadly. I don't have a ready suggestion here.

Reviewer 2: ```suggestion
The `semgrep-ci` hook requires the user to be logged in locally and runs with the engine configured in the organization, but the standard `semgrep` hook can also take advantage of local login to run with Pro rules and cross-function analysis.
```
Does this work?

Author: yeah, I think that's fine and probably clearer than Semgrep Code! Thanks!
